src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1603,6 +1613,49 @@ public override BoundNode VisitCall(BoundCall node)
             return null;
         }
 
+        /// <summary>
+        /// For each argument, figure out if its corresponding parameter is annotated with NotNullWhenFalse or
+        /// EnsuresNotNull.
+        /// </summary>
+        private static ImmutableArray<AttributeAnnotations> GetAnnotations(int numArguments,
+            bool expanded, ImmutableArray<ParameterSymbol> parameters, ImmutableArray<int> argsToParamsOpt)
+        {
+            ArrayBuilder<AttributeAnnotations> builder = null;
+
+            for (int i = 0; i < numArguments; i++)
+            {
+                (ParameterSymbol parameter, _) = GetCorrespondingParameter(i, parameters, argsToParamsOpt, expanded);
+                AttributeAnnotations annotations = parameter?.FlowAnalysisAnnotations ?? AttributeAnnotations.None;
+
+                // We'll ignore NotNullWhenFalse that is misused in metadata
+                if ((annotations & AttributeAnnotations.NotNullWhenFalse) != 0 &&
+                    parameter.ContainingSymbol.GetTypeOrReturnType().SpecialType != SpecialType.System_Boolean)
```

Comment:
>parameter [](start = 20, length = 9)

Should we use original definition for this check? Is generic case covered by tests?

Reply:
Added ` NotNullWhenFalse_RequiresBoolReturn_OnGenericMethod`. The behavior of checking the specialized definition, rather than the original definition, seem ok.

---
In reply to: [187948468](https://github.com/dotnet/roslyn/pull/26656#discussion_r187948468) [](ancestors = 187948468)